lib/kubernetes-deploy/deploy_task.rb

```diff
@@ -475,13 +475,23 @@ def record_apply_failure(err, resources: [])
         .select(&:sensitive_template_content?)
         .map { |r| File.basename(r.file_path) }
 
+      server_dry_run_validated_resource = resources
+        .select(&:server_dry_run_validated?)
+        .map { |r| File.basename(r.file_path) }
+
       err.each_line do |line|
         bad_files = find_bad_files_from_kubectl_output(line)
         if bad_files.present?
           bad_files.each do |f|
             if filenames_with_sensitive_content.include?(f[:filename])
-              # Hide the error and template contents in case it has senitive information
-              record_invalid_template(err: "SUPPRESSED FOR SECURITY", filename: f[:filename], content: nil)
+              # Hide the error and template contents in case it has sensitive information
```

Comment:
I don't think we should pass the file content. I don't think I've ever seen it triggered so I think being overly cautions here is ok. My main concern is that in the future there is a new sensitive type that we correctly know is sensitive but doesn't match the regexp check `/kind:\s*Secret/`. So `record_invalid_template` blindly dumps the file to logs. 

I think passing the err is fine since the entire PR is based on the idea that with `server-dry-run` the errors at the apply stage wont contain sensitive values.

Reply:
Yes, I think we are good to pass the error if resource is able to go through `server-dry-run`.

About the concern in `record_invalid_template`, I feel the reason we have `/kind:\s*Secret/` is we worry there's resource who has `/kind:\s*Secret/` but we didn't recognize it as `secret`. I kind of think it's over caution.

Reply:
> I feel the reason we have /kind:\s*Secret/ is we worry there's resource who has /kind:\s*Secret/ but we didn't recognize it as secret. 

IIRC the regex is there because `record_invalid_template` is used for errors occurring during rendering too, and at that point it might be possible for the resource to have been completely invalid and therefore not recognized as anything in particular. The `kind: Secret` scan is a last-ditch effort to prevent leaks when our more robust upstream checks failed. Is it necessary? I hope not. But since there is in reality only this one sensitive resource, I believe the consensus when it went it was that it doesn't hurt.

Reply:
More importantly, didn't we agree to trust kubectl output for the error message itself (and just suppress the template content)? https://github.com/Shopify/kubernetes-deploy/issues/484#issuecomment-496594881

Reply:
`record_invalid_template` is also used during rendering, yeah, agree that `/kind:\sSecret/ ` is a last-ditch effort we need.  

> didn't we agree to trust kubectl output for the error message itself (and just suppress the template content)? #484 (comment)

We did an experiment and raise issue https://github.com/kubernetes/kubernetes/issues/82421, it looks we can error message after server dry-run. Did I understand your comment correctly?

Reply:
In issue 484, I identified three related problems:
1- was kubernetes-deploy right to fail this deploy?
2-the error was censored, and didn't need to be
3-secret keys are subject to a validation that dry run isn't catching

Server-side dry run addresses problem 3. What I'm talking about is our conclusion on problem 2, which I think would considerably simplify your code here. tl;dr Danny and I concluded on the issue that upstream is being careful about what they put in the error message they give us, and suppressing the error message was overzealous on our part.

Reply:
Never mind -- I missed the part of the issue you opened that _proved_ that apply can dump secret content at least in one specific case. That's enough to warrant what you're doing here. 👍

Reply:
Yeah, the apply error message does leak secret content unfortunately. We can only trust output after server-dry-run for now.